File: workbench/security_boundary_demo.py
"""Executable audit of current MIN0 CORE FORTH security controls and gaps."""

from __future__ import annotations

import copy
import json

from constructor_image_fixture import make_system
from min0_core_forth_image import (
    ImageError,
    build_image_envelope,
    validate_image_envelope,
)
from min0_core_forth_outer import OuterInterpreter, install_core_primitives
from min0_core_forth_vm import StepLimitExceeded
from image_envelope_demo import (
    SOURCE_BASES,
    SOURCE_LIMITS,
    build_source_image,
)


def _scenario(threat_id: str, name: str, result: str, status: str) -> dict[str, str]:
    return {"id": threat_id, "scenario": name, "result": result, "status": status}


def run_demo(implementation: str = "python") -> dict:
    components, envelope = build_source_image()

    corrupted = dict(components)
    corrupted_code = bytearray(corrupted["code"])
    corrupted_code[-1] = 0  # Valid NOP: altered semantics, still valid bytecode.
    corrupted["code"] = bytes(corrupted_code)
    try:
        validate_image_envelope(corrupted, envelope)
    except ImageError:
        corruption = "blocked"
    else:
        corruption = "accepted"

    manifest_tamper = copy.deepcopy(envelope)
    manifest_tamper["manifest"]["records"][0]["kind"] += "-tampered"
    try:
        validate_image_envelope(components, manifest_tamper)
    except ImageError:
        manifest_result = "blocked"
    else:
        manifest_result = "accepted"

    rebuilt_envelope = build_image_envelope(
        corrupted,
        SOURCE_BASES,
        SOURCE_LIMITS,
        envelope["allocator"],
        envelope["manifest"],
        generation=envelope["generation"],
    )
    validate_image_envelope(corrupted, rebuilt_envelope)
    malicious_development = "accepted"
    try:
        validate_image_envelope(
            corrupted, rebuilt_envelope, require_authentication=True
        )
    except ImageError:
        malicious_authenticated = "blocked"
    else:
        malicious_authenticated = "accepted"

    old_components, old_envelope = build_source_image(envelope["generation"] - 1)
    try:
        validate_image_envelope(
            old_components,
            old_envelope,
            minimum_generation=envelope["generation"],
        )
    except ImageError:
        rollback_result = "blocked"
    else:
        rollback_result = "accepted"

    vm, dictionary = make_system()
    install_core_primitives(dictionary)
    outer = OuterInterpreter(vm, dictionary)
    outer.interpret(": FOREVER BEGIN AGAIN ;")
    forever = dictionary.find("FOREVER")
    assert forever is not None
    try:
        vm.resume(forever.payload, return_to=outer.return_trampoline, max_steps=20)
    except StepLimitExceeded:
        execution_limit = "blocked"
    else:
        execution_limit = "accepted"

    scenarios = [
        _scenario("T01", "component-corruption", corruption, "controlled"),
        _scenario("T02", "manifest-tamper", manifest_result, "controlled"),
        _scenario(
            "T03", "malicious-rebuild-development", malicious_development, "gap"
        ),
        _scenario(
            "T04",
            "malicious-rebuild-authentication-required",
            malicious_authenticated,
            "policy-boundary",
        ),
        _scenario("T05", "rollback-old-valid-image", rollback_result, "controlled"),
        _scenario("T06", "infinite-execution", execution_limit, "controlled"),
    ]
    return {
        "implementation": implementation,
        "authentication": envelope["authentication"]["scheme"],
        "generation_present": "generation" in envelope,
        "scenarios": scenarios,
        "controlled": sum(item["status"] == "controlled" for item in scenarios),
        "gaps": [item["id"] for item in scenarios if item["status"] == "gap"],
    }


if __name__ == "__main__":
    print(json.dumps(run_demo(), sort_keys=True, separators=(",", ":")))

File: workbench/security_boundary_demo.rb
# frozen_string_literal: true

require "json"
require_relative "image_envelope_demo"

def security_scenario(threat_id, name, result, status)
  { id: threat_id, scenario: name, result: result, status: status }
end

def run_security_boundary_demo(implementation = "ruby")
  components, envelope = build_source_image

  corrupted = components.transform_values(&:dup)
  last = corrupted[:code].bytesize - 1
  corrupted[:code].setbyte(last, 0) # Valid NOP: altered semantics, still valid bytecode.
  corruption = begin
    Min0CoreForth::ImageEnvelope.validate(corrupted, envelope)
    "accepted"
  rescue Min0CoreForth::ImageError
    "blocked"
  end

  manifest_tamper = Marshal.load(Marshal.dump(envelope))
  manifest_tamper[:manifest][:records][0][:kind] += "-tampered"
  manifest_result = begin
    Min0CoreForth::ImageEnvelope.validate(components, manifest_tamper)
    "accepted"
  rescue Min0CoreForth::ImageError
    "blocked"
  end

  rebuilt_envelope = Min0CoreForth::ImageEnvelope.build(
    corrupted,
    SOURCE_BASES,
    SOURCE_LIMITS,
    envelope[:allocator],
    envelope[:manifest],
    generation: envelope[:generation]
  )
  Min0CoreForth::ImageEnvelope.validate(corrupted, rebuilt_envelope)
  malicious_development = "accepted"
  malicious_authenticated = begin
    Min0CoreForth::ImageEnvelope.validate(
      corrupted, rebuilt_envelope, require_authentication: true
    )
    "accepted"
  rescue Min0CoreForth::ImageError
    "blocked"
  end

  old_components, old_envelope = build_source_image(envelope[:generation] - 1)
  rollback_result = begin
    Min0CoreForth::ImageEnvelope.validate(
      old_components, old_envelope, minimum_generation: envelope[:generation]
    )
    "accepted"
  rescue Min0CoreForth::ImageError
    "blocked"
  end

  vm, dictionary = make_image_system
  Min0CoreForth.install_core_primitives(dictionary)
  outer = Min0CoreForth::OuterInterpreter.new(vm, dictionary)
  outer.interpret(": FOREVER BEGIN AGAIN ;")
  forever = dictionary.find("FOREVER")
  execution_limit = begin
    vm.resume(forever.payload, return_to: outer.return_trampoline, max_steps: 20)
    "accepted"
  rescue Min0CoreForth::StepLimitExceeded
    "blocked"
  end

  scenarios = [
    security_scenario("T01", "component-corruption", corruption, "controlled"),
    security_scenario("T02", "manifest-tamper", manifest_result, "controlled"),
    security_scenario("T03", "malicious-rebuild-development", malicious_development, "gap"),
    security_scenario(
      "T04", "malicious-rebuild-authentication-required", malicious_authenticated, "policy-boundary"
    ),
    security_scenario("T05", "rollback-old-valid-image", rollback_result, "controlled"),
    security_scenario("T06", "infinite-execution", execution_limit, "controlled")
  ]
  {
    implementation: implementation,
    authentication: envelope[:authentication][:scheme],
    generation_present: envelope.key?(:generation),
    scenarios: scenarios,
    controlled: scenarios.count { |item| item[:status] == "controlled" },
    gaps: scenarios.select { |item| item[:status] == "gap" }.map { |item| item[:id] }
  }
end

puts JSON.generate(run_security_boundary_demo) if $PROGRAM_NAME == __FILE__
